### DISCUSSION TITLE
Add Permissions/Roles Changed/Revoked events

### DISCUSSION THREAD

**Discussion by ghost** [spatie/laravel-permission#1943] [Ideas]
Edit: 

The original idea for a pivot model stemmed from having more control. I've reformulated this using an event [here](#discussioncomment-1715846)

Original:

----

> Awesome package, one minor addition which would be nice:
> 
> Using a default pivot model class to the BelongsToMany relation methods (role and permission) will allow for the ability to attach an observer to them.
> 
> The use-case could be plenty, what comes to mind with is retracting security tokens, logging users out, etc.
> 
> What I have done is something like this in my own model, but an out of the box endpoint would be nice.
> 
> ```
> public function permissions(): BelongsToMany
>     {
>         return $this->morphToMany(
>             config('permission.models.permission'),
>             'model',
>             config('permission.table_names.model_has_permissions'),
>             config('permission.column_names.model_morph_key'),
>             'permission_id'
>         )->using(DefaultPivot::class);
>     }
> ```

**Comment by erikn69** [comment]
Maybe better
```php
public function permissions(): BelongsToMany
{
    return parent::permissions()->using(DefaultPivot::class);
}
```
Authors wants to keep the package simple

**Comment by ghost** [comment]
Went back to this just now.. I realize overwriting the trait method `public function syncPermissions(...$permissions)` is the easiest way (the same for HasRoles).

This method detaches everything first, and follows up with permission/role assignment.

An event here would still be nice to keep things clean imo. 

Something like

```php
public function syncPermissions(...$permissions)
{
    $oldPermissions = $this->permissions->pluck('name')->flatten();
    $newPermissions = collect($permissions)->flatten();

    $this->permissions()->detach();
    $this->givePermissionTo($permissions);

    PermissionChangeEvent::dispatch($oldPermissions, $newPermissions);

    // Or maybe more usefull
    // $revoked = $oldPermissions->diff($newPermissions);
    // if ($revoked->isNotEmpty()) {
    //     PermissionsRevokedEvent::dispatch($revoked);
    // }

    return $this;
}
```

Nothing too fancy or complicated.

Anyway, it's also solvable without. Feel free to close if there is no interest.
